utils/pstor/src/etcd_watcher.rs

```diff
@@ -0,0 +1,497 @@
+    fn handle_request_idle(&mut self, op: Option<WatchRequest<Ctx>>) -> WatchState {
+        match op {
+            None => WatchState::Closed,
+            Some(WatchRequest::Register(register)) => {
+                self.config.register(register);
+                WatchState::Connecting
+            }
+            Some(WatchRequest::Unregister { prefix }) => {
+                self.config.unregister(&prefix);
+                WatchState::Idle
+            }
+        }
+    }
+
+    /// Handles a new [`WatchRequest`] when we are already in the [`WatchState::Watching`] state.
+    /// In this state, we may receive new requests, and we also process watch events.
+    async fn handle_request(
+        &mut self,
+        mut watcher: StreamedWatcher,
+        request: Option<WatchRequest<Ctx>>,
+    ) -> WatchState {
+        match request {
+            None => {
+                watcher.watcher.cancel().await.ok();
+                WatchState::Closed
+            }
+            Some(WatchRequest::Register(register)) => {
+                let prefix = register.prefix.clone();
+                let rev = register.rev;
+                self.config.register(register);
+                match watcher.watch_key_prefix(prefix, rev).await {
+                    Ok(_) => WatchState::Watching(watcher),
+                    Err(error) => self.handle_error(&error, "Failed to watch the given key"),
+                }
+            }
+            Some(WatchRequest::Unregister { prefix }) => {
+                self.config.unregister(&prefix);
+                watcher.del_watch_key_prefix(&prefix).await;
+                WatchState::Watching(watcher)
+            }
+        }
+    }
+
+    /// When an error is encountered we go into the [`WatchState::Reconnecting`] state where we
+    /// attempt to re-establish the watch connection stream.
+    async fn handle_reconnecting(
+        &mut self,
+        mut attempts: u32,
+        request: &mut tokio::sync::mpsc::UnboundedReceiver<WatchRequest<Ctx>>,
+    ) -> WatchState {
+        if attempts % 10 == 0 {
+            tracing::warn!("Starting etcd watch stream reconnection loop, attempt: {attempts}");
+        }
+
+        // ensure we keep processing incoming requests, even during signal loss
+        self.drip_pending_requests(request);
+
+        let delay = (Duration::from_secs(1) * attempts).min(Duration::from_secs(5));
+        attempts += 1;
+        tokio::time::sleep(delay).await;
+        match self.handle_connecting().await {
+            WatchState::Reconnecting(_) => WatchState::Reconnecting(attempts),
+            WatchState::Watching(w) => {
+                tracing::info!("Successfully re-connected the watch stream");
+                WatchState::Watching(w)
+            }
+            state => state,
+        }
+    }
+
+    /// When the connection is lost and we're attempting to reconnect, ensure the requests keep
+    /// getting processed.
+    fn drip_pending_requests(
```

Comment:
drip --> drop?

Reply:
hmm it's not dropping them right? It's processing some of the pending request to avoid starving them out.
So I thought of drip as it's dripping them :D 
Any other suggestion?

Reply:
Okay, yeah drip sounds right in that case.